### PR TITLE
Warn the user in the crafting menu when canning rotten ingredients

### LINKED ISSUE #38271: Cooking canned meat from rotten scraps of meat doesn't get marked as using rotten ingredients
# Describe the bug

And correctly produces a can of rotten meat.

# Steps To Reproduce

Load the game, make canned meat.

# Expected behavior

Canned meat getting the label and the rotten color in the crafting menu as other meat recipes.

# Screenshots

# Versions and configuration
- OS: Linux
    - OS Version: <unknown>
- Game Version: 0.D-12504-g90156182f7-dirty [64-bit]
- Graphics Version: Curses
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Alternative Map Key [alt_map_key],
    No Rail Stations [No_Rail_Stations]
]

# Additional context

[2020.02.22-1.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/4241815/2020.02.22-1.zip)

### PR DESCRIPTION
**Pull request by jkraybill**
#### Summary
```SUMMARY: Bugfixes "warn the user in the crafting menu when canning rotten ingredients"```

#### Purpose of change
Fixes #38271.

The issue was that the filter for the interface was checking if the resulting item goes_bad(), which preserved items do not return true for.

So I added a new check to item that returns true if it goes bad, OR if it is a preserving container whose contents go bad.

#### Describe the solution
I added a new check to item.cpp which returns true if this item goes bad, or if this item is a preserving container and its contents can go bad.

#### Describe alternatives you've considered
I initially just added the container checks to recipe.cpp, but that seemed like bad encapsulation.

I called the new method "goes_bad_after_opening" and considered other names; the current name probably isn't properly representing that it applies to items whether or not they are contained.

#### Testing
Tested different stacks of meat and fruit, rotten and fresh, and canning them before and afer the change.

This screenshot shows that after the change, canned recipes using rotten ingredients are properly highlighted:
sc001358

In this screenshot, there were a small number of fresh ingredients and a large number of rotten ones, to confirm that the recipes conditionally show the right condition under these circumstances (the brown recipes are the ones that use large ingredient quanitities so would only use the rotten ingredients):
sc001359

#### Additional context
As with my previous commits, I'm still no expert in the codebase and rusty in C++, so if there is a better way to do what I am doing here, please let me know!

### CODE REVIEW

**Comment by Amariithynar** [comment]
Some items will actually gain lifespan when processed, like making fruit juice from rose hips or making condensed or fermenting curdling milk made from rotten milk, as the raw product will go bad but can still be processed using various methods to get perfectly usable materials from them. Their lifespans are usually measured in just a few hours, at best, when this is done, though, which seems fine to me; basically, use it up immediately, or throw it out.

Also of note is that if there are enough resources to do a recipe even a single time, it won't show up as using rotten ingredients, until you try to [b]atch them, at which point any batches that use rotten ingredients will be listed in brown. Most likely if you enter the batching screen for any of those, you'll see one or two white batches, and the rest in brown (or grey).

File: src/item.cpp [hunk 1]
```diff
@@ -4740,6 +4740,13 @@ bool item::goes_bad() const
     return is_food() && get_comestible()->spoils != 0_turns;
 }
 
+bool item::goes_bad_after_opening() const
+{
+    // returns true if this item goes bad, or if this item is a preserving container and its contents can go bad.
+    return goes_bad() || ( type->container && type->container->preserves &&
+                           contents.front().goes_bad() );
```
Comment (BevapDin) [inline src/item.cpp]: You may want to check that `contents` actually contains something before accessing its contents.

File: src/item.cpp [hunk 2]
```diff
@@ -4740,6 +4740,13 @@ bool item::goes_bad() const
     return is_food() && get_comestible()->spoils != 0_turns;
 }
 
+bool item::goes_bad_after_opening() const
+{
+    // returns true if this item goes bad, or if this item is a preserving container and its contents can go bad.
```
Comment (BevapDin) [inline src/item.cpp]: There is already documentation for this in the header. Why repeat this here? And why repeat the code as comment? The comment doesn't tell more than the code itself already does.

**Comment by jkraybill** [comment]
@Amariithynar can you clarify what you mean about fruit juice and milk? Is there a bug there?

@Bevapdin thanks for the notes, will address those.

File: src/item.cpp (hunk 2 above)
Comment (jkraybill) [reply src/item.cpp]: Fixed.

File: src/item.cpp (hunk 1 above)
Comment (jkraybill) [reply src/item.cpp]: Done.

**Comment by jkraybill** [comment]
That failure is that nagging one that isn't from this PR:

```
explosion_balance_test.cpp:130: FAILED:
  CHECK( before_hp[ i ] == after_hp[ i ] )
with expansion:
  120 == 108
with message:
  <color_c_yellow>|\</color> car battery
```

**Comment by Amariithynar** [comment]
No bug, though there's probably some that would argue there is.

Rose Hips have an expiry of only 2 days or so, but you can squeeze them for fruit juice after they've already reached "rotten" and the fruit juice will be past its midlife, but still usable (I'm actually surprised they're set to so few days, and there's no way to dry them for rose hip tea; rose hips last for a fairly long time).

Similarly, you can take "rotten" milk and curdle it, then ferment the "rotten" curdling milk to get edible (but past its midlife) curdled milk, which you can process into hard cheese, which is relatively accurate to real life.

Milk does not actually *rot* very easily, but instead tends to sour rapidly, making it very unpalatable to our tastebuds, but still edible; Game mechanics don't really have an equivalent, so it just is "rotten"; milk left for a long time in a jug, even the hyperpasteurized stuff sold in stores, would more likely have turned into a thick yogurt with a thin whey, if left sealed, or into soured milk if opened and then resealed. Soured milk can still be processed into cheese with minimal issue, and you can actually use yogurt as a starter culture to do so. It's a bit odd that you *consume* a stomach to make it, though, instead of either a) cultivate rennet from inside a stomach bladder, as we do with yeast (you can use whey from previously making cheese as a rennet starter similar how you can use a sourdough starter to make more; yogurt can be used naturally for that) or b) use the stomach as a storage container to ferment the milk into curdling milk (how it was first discovered, historically, as stomachs were used as waterskins). You can also use a juice high in citric acid as an acidifier (I've made a few mozzarellas using powdered citric acid, orange juice, and lemon juice, for example) instead of relying on rennet, too.

Similarly, you can take "rotten" milk and cook it down with sugar to make condensed milk that won't last much longer, but has been rendered edible once again for a short period longer; similar to how many meats were smothered in spices to mask the taste change from it starting to go off. The canning process also helps re-sterilize it, so this makes sense. It probably should still have a flag for "tastes funky" or similar.

**Comment by jkraybill** [comment]
@Amariithynar thanks for the notes! Definitely "food for thought" for some post-0.E food updates.

**Pull request merged**